integration/pkcs11/pkcs11_test.go

```diff
@@ -0,0 +1,411 @@
+	caKey := key.(*ecdsa.PrivateKey)
+
+	certTemplate := &x509.Certificate{
+		Signature:          csr.Signature,
+		SignatureAlgorithm: csr.SignatureAlgorithm,
+		PublicKey:          csr.PublicKey,
+		PublicKeyAlgorithm: csr.PublicKeyAlgorithm,
+
+		SerialNumber:          serialNumber,
+		NotBefore:             time.Now().Add(-1 * time.Minute).UTC(),
+		NotAfter:              time.Now().Add(365 * 24 * time.Hour).UTC(),
+		BasicConstraintsValid: true,
+
+		Subject:     csr.Subject,
+		KeyUsage:    x509.KeyUsageDigitalSignature,
+		ExtKeyUsage: []x509.ExtKeyUsage{},
+	}
+
+	// Use root CA to create and sign cert
+	signedCert, err := x509.CreateCertificate(rand.Reader, certTemplate, caCert, pubKey, caKey)
+	Expect(err).ToNot(HaveOccurred())
+
+	return pem.EncodeToMemory(&pem.Block{Bytes: signedCert, Type: "CERTIFICATE"})
+}
+
+// Overwrites existing cert and removes private key from keystore folder
+func updateMSPFolder(path, certName string, cert []byte) {
+	// Overwrite existing certificate with new certificate
+	signCertsPath := filepath.Join(path, "signcerts")
+	err := ioutil.WriteFile(fmt.Sprintf("%s/%s", signCertsPath, certName), cert, 0644)
+	Expect(err).NotTo(HaveOccurred())
+
+	// delete the existing private key - this is stored in the hsm
+	adminKSCert := filepath.Join(path, "keystore", "priv_sk")
+	err = os.Remove(adminKSCert)
+	Expect(err).NotTo(HaveOccurred())
+}
+
+// Generating key pair in HSM, convert, and return keys
+func generateKeyPair(ctx *pkcs11.Ctx, sess pkcs11.SessionHandle) (*ecdsa.PublicKey, *P11ECDSAKey) {
+	publabel, privlabel := "BCPUB7", "BCPRV7"
+
+	curve := asn1.ObjectIdentifier{1, 2, 840, 10045, 3, 1, 7} // secp256r1 Curve
+
+	marshaledOID, err := asn1.Marshal(curve)
+	Expect(err).ToNot(HaveOccurred())
+
+	pubAttrs := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, pkcs11.CKK_EC),
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PUBLIC_KEY),
+		pkcs11.NewAttribute(pkcs11.CKA_TOKEN, true),
+		pkcs11.NewAttribute(pkcs11.CKA_VERIFY, true),
+		pkcs11.NewAttribute(pkcs11.CKA_EC_PARAMS, marshaledOID),
+
+		pkcs11.NewAttribute(pkcs11.CKA_ID, publabel),
+		pkcs11.NewAttribute(pkcs11.CKA_LABEL, publabel),
+	}
+
+	privAttrs := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, pkcs11.CKK_EC),
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PRIVATE_KEY),
+		pkcs11.NewAttribute(pkcs11.CKA_TOKEN, true),
+		pkcs11.NewAttribute(pkcs11.CKA_SIGN, true),
+
+		pkcs11.NewAttribute(pkcs11.CKA_ID, privlabel),
+		pkcs11.NewAttribute(pkcs11.CKA_LABEL, privlabel),
+
+		pkcs11.NewAttribute(pkcs11.CKA_EXTRACTABLE, false),
+		pkcs11.NewAttribute(pkcs11.CKA_SENSITIVE, true),
+	}
+
+	pubK, privK, err := ctx.GenerateKeyPair(
+		sess,
+		[]*pkcs11.Mechanism{pkcs11.NewMechanism(pkcs11.CKM_EC_KEY_PAIR_GEN, nil)},
+		pubAttrs,
+		privAttrs,
+	)
+	Expect(err).ToNot(HaveOccurred())
+
+	ecpt := ecPoint(ctx, sess, pubK)
+	Expect(err).NotTo(HaveOccurred())
+
+	hash := sha256.Sum256(ecpt)
+	ski := hash[:]
```

Comment:
`ski := sha256.Sum256(ecpt)[:]`

Reply:
I don't think arrays can be sliced into until they're actually allocated, the compiler will throw a `slice of unaddressable value` error if you try to one-line it. https://github.com/golang/go/blob/21ec72c2ca5168f3f10b4594a553b3a038c8df29/test/complit1.go

Reply:
I can't argue with the compiler.